File: packages/ink-swift/Tests/InkCalligrapherTests/EngineTests.swift
/// End-to-end calligrapher engine behavior, mirroring the spirit of the
/// graves engine tests.

import Foundation
import InkCalligrapher
import InkCore
import Testing

@Suite struct EngineTests {
    let model = try! CalligrapherModel(assets: Fixtures.assets)

    @Test func writesALineAndTerminatesNaturally() throws {
        let offsets = try model.write("hello world", seed: 42)
        #expect(offsets.count > 50)
        #expect(offsets.count <= 40 * "hello world".count)
        #expect(offsets.allSatisfy { $0.dx.isFinite && $0.dy.isFinite })
        #expect(offsets.filter(\.eos).count > 3)
    }

    @Test func isDeterministicPerSeedIncludingRandomStyle() throws {
        // style nil draws the style from the seed, so it must reproduce too.
        let first = try model.write("hello", seed: 7)
        let second = try model.write("hello", seed: 7)
        let different = try model.write("hello", seed: 8)
        #expect(first == second)
        #expect(first != different)
    }

    @Test func encodesWithMarkersAndUnknownFallback() {
        #expect(model.encode("hi") == [2, 30, 13, 3])
        #expect(model.encode("h~i") == [2, 30, 1, 13, 3])
        #expect(model.supports("h"))
        #expect(!model.supports("~"))
        #expect(model.styles.count == 80)
    }

    @Test func rejectsUnknownStyles() {
        #expect(throws: CalligrapherError.self) {
            try model.writer("hi", style: 99)
        }
    }

    /// Real-time streaming needs 125 steps/sec (the web app reveals one
    /// step per 8ms). Debug builds run -Onone (no unsafeFlags in a
    /// remotely-consumable manifest), which misses the bar by design, so
    /// the gate only exists in release runs: `swift test -c release`.
    #if !DEBUG
    @Test func sustainsAtLeast125StepsPerSecond() throws {
        let writer = try model.writer(
            "the quick brown fox jumps over the lazy dog then keeps on going",
            bias: 0.75,
            style: 3,
            seed: 1
        )
        for _ in 0 ..< 30 { _ = writer.step() } // warmup
        let clock = ContinuousClock()
        var steps = 0
        let elapsed = clock.measure {
            while steps < 600, writer.step() != nil { steps += 1 }
        }
        let stepsPerSecond = Double(steps) / max(elapsed.seconds, 1e-9)
        print("engine speed: \(Int(stepsPerSecond)) steps/sec over \(steps) steps")
        #expect(stepsPerSecond > 125)
    }
    #endif
}

private extension Duration {
    var seconds: Double {
        let (seconds, attoseconds) = components
        return Double(seconds) + Double(attoseconds) * 1e-18
    }
}

File: packages/ink-kotlin/ink-calligrapher/src/test/kotlin/com/trylonghand/ink/calligrapher/EngineTests.kt
/// End-to-end calligrapher engine behavior, mirroring the spirit of the
/// graves engine tests.

package com.trylonghand.ink.calligrapher

import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class EngineTests {
    private val model = CalligrapherModel(Fixtures.assets)

    @Test
    fun writesALineAndTerminatesNaturally() {
        val offsets = model.write("hello world", seed = 42u)
        assertTrue(offsets.size > 50)
        assertTrue(offsets.size <= 40 * "hello world".length)
        assertTrue(offsets.all { it.dx.isFinite() && it.dy.isFinite() })
        assertTrue(offsets.count { it.eos } > 3)
    }

    @Test
    fun isDeterministicPerSeedIncludingRandomStyle() {
        // style null draws the style from the seed, so it must reproduce too.
        val first = model.write("hello", seed = 7u)
        val second = model.write("hello", seed = 7u)
        val different = model.write("hello", seed = 8u)
        assertEquals(first, second)
        assertNotEquals(first, different)
    }

    @Test
    fun encodesWithMarkersAndUnknownFallback() {
        assertContentEquals(intArrayOf(2, 30, 13, 3), model.encode("hi"))
        assertContentEquals(intArrayOf(2, 30, 1, 13, 3), model.encode("h~i"))
        assertTrue(model.supports('h'))
        assertFalse(model.supports('~'))
        assertEquals(80, model.styles.size)
    }

    @Test
    fun rejectsUnknownStyles() {
        assertFailsWith<CalligrapherError> {
            model.writer("hi", style = 99)
        }
    }

    /**
     * Real-time streaming needs 125 steps/sec (the web app reveals one
     * step per 8ms). The Swift port gates this in release builds only; on
     * the JVM the JIT-compiled test run is the release analog, so the gate
     * always runs here.
     */
    @Test
    fun sustainsAtLeast125StepsPerSecond() {
        val writer = model.writer(
            "the quick brown fox jumps over the lazy dog then keeps on going",
            bias = 0.75,
            style = 3,
            seed = 1u,
        )
        repeat(30) { writer.step() } // warmup
        var steps = 0
        val start = System.nanoTime()
        while (steps < 600 && writer.step() != null) steps += 1
        val elapsed = (System.nanoTime() - start) * 1e-9
        val stepsPerSecond = steps.toDouble() / maxOf(elapsed, 1e-9)
        println("engine speed: ${stepsPerSecond.toInt()} steps/sec over $steps steps")
        assertTrue(stepsPerSecond > 125)
    }
}
